Segfault with *** buffer overflow detected ***: utox terminated;

Overflow in __builtin___memcpy_chk in /usr/include/bits/string3.h:53:3

Build from https://github.com/uTox/uTox/commit/35db020ba872f45af0c54cd0ff92649b8598d7fd produces a warning:
```
[ 76%] Building C object CMakeFiles/utox.dir/src/groups.c.o
/usr/lib64/ccache/bin/x86_64-pc-linux-gnu-gcc   -I/usr/include/freetype2   -DNDEBUG -march=core2 -mmmx -mno-3dnow -msse -msse2 -msse3 -mssse3 -mno-sse4a -mcx16 -msahf -mno-movbe -mno-aes -mno-sha -mno-pclmul -mno-popcnt -mno-abm -mno-lwp -mno-fma -mno-fma4 -mno-xop -mno-bmi -mno-bmi2 -mno-tbm -mno-avx -mno-avx2 -mno-sse4.2 -msse4.1 -mno-lzcnt -mno-rtm -mno-hle -mno-rdrnd -mno-f16c -mno-fsgsbase -mno-rdseed -mno-prfchw -mno-adx -mfxsr -mno-xsave -mno-xsaveopt -mno-avx512f -mno-avx512er -mno-avx512cd -mno-avx512pf -mno-prefetchwt1 --param l1-cache-size=32 --param l1-cache-line-size=64 --param l2-cache-size=3072 -mtune=core2 -O2 -pipe  -Wall -Wextra -Werror=implicit-function-declaration -Wformat=0 -fno-strict-aliasing -DAUDIO_FILTERING   -std=gnu11 -o CMakeFiles/utox.dir/src/groups.c.o -c /var/tmp/portage/net-im/utox-9999/work/utox-9999/src/groups.c
/var/tmp/portage/net-im/utox-9999/work/utox-9999/src/friend.c: In function ‘friend_meta_data_read’:
/var/tmp/portage/net-im/utox-9999/work/utox-9999/src/friend.c:84:5: warning: ignoring return value of ‘fread’, declared with attribute warn_unused_result [-Wunused-result]
     fread(metadata, size, 1, file);
     ^
In file included from /usr/include/string.h:635:0,
                 from /usr/include/freetype2/freetype/config/ftstdlib.h:77,
                 from /usr/include/freetype2/freetype/config/ftconfig.h:43,
                 from /usr/include/freetype2/freetype/freetype.h:33,
                 from /usr/include/freetype2/freetype/ftlcdfil.h:24,
                 from /var/tmp/portage/net-im/utox-9999/work/utox-9999/src/xlib/freetype.h:9,
                 from /var/tmp/portage/net-im/utox-9999/work/utox-9999/src/xlib/main.h:13,
                 from /var/tmp/portage/net-im/utox-9999/work/utox-9999/src/main_native.h:11,
                 from /var/tmp/portage/net-im/utox-9999/work/utox-9999/src/groups.c:6:
In function ‘memcpy’,
    inlined from ‘group_peer_add’ at /var/tmp/portage/net-im/utox-9999/work/utox-9999/src/groups.c:88:6:
/usr/include/bits/string3.h:53:3: warning: call to __builtin___memcpy_chk will always overflow destination buffer
   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
   ^
```
https://github.com/robinlinden/uTox/commit/516e62d5db28a96493a17bccddaa74b335952082 workaround resolves the issue

--- a/src/groups.c
+++ b/src/groups.c
@@ -84,7 +84,8 @@ void group_peer_add(GROUPCHAT *g, uint32_t peer_id, bool UNUSED(our_peer_number)
 
     GROUP_PEER *peer = g->peer[peer_id];
 
-    peer = calloc(1, sizeof(GROUP_PEER));
+    // Allocate space for the struct and the dynamic array holding the peer's name.
+    peer = calloc(1, sizeof(GROUP_PEER) + 10); // 10 is strlen("<unknown>") + \0
     strcpy2(peer->name, "<unknown>");
     peer->name_length = 0;
     peer->name_color  = name_color;

--- a/src/groups.c
+++ b/src/groups.c
@@ -68,6 +68,16 @@ uint32_t group_add_message(GROUPCHAT *g, uint32_t peer_id, const uint8_t *messag
     msg->via.grp.author_color  = peer->name_color;
     time(&msg->time);
 
+    msg->via.grp.author = calloc(1, peer->name_length);
+    if (!msg->via.grp.author) {
+        debug_error("Groupchat:\tUnable to allocate space for author nickname.\n");
+    }
+
+    msg->via.grp.msg = calloc(1, length);
+    if (!msg->via.grp.msg) {
+        debug_error("Groupchat:\tUnable to allocate space for message.\n");
+    }
+
     memcpy(msg->via.grp.author, nick,    peer->name_length);
     memcpy(msg->via.grp.msg,    message, length);
 

--- a/src/groups.c
+++ b/src/groups.c
@@ -54,16 +54,16 @@ void group_init(GROUPCHAT *g, uint32_t group_number, bool av_group) {
 
 uint32_t group_add_message(GROUPCHAT *g, uint32_t peer_id, const uint8_t *message, size_t length, uint8_t m_type) {
     pthread_mutex_lock(&messages_lock); /* make sure that messages has posted before we continue */
-    MESSAGES   *m    = &g->msg;
-    GROUP_PEER *peer = g->peer[peer_id];
-    uint8_t    *nick = peer->name;
+    const GROUP_PEER *peer = g->peer[peer_id];
+    MESSAGES *m = &g->msg;
 
     MSG_HEADER *msg    = calloc(1, sizeof(MSG_HEADER));
-    msg->our_msg       = (g->our_peer_number == peer_id ? 1 : 0);
+    msg->our_msg       = (g->our_peer_number == peer_id ? true : false);
     msg->msg_type      = m_type;
 
     msg->via.grp.length        = length;
     msg->via.grp.author_id     = peer_id;
+
     msg->via.grp.author_length = peer->name_length;
     msg->via.grp.author_color  = peer->name_color;
     time(&msg->time);
@@ -71,17 +71,24 @@ uint32_t group_add_message(GROUPCHAT *g, uint32_t peer_id, const uint8_t *messag
     msg->via.grp.author = calloc(1, peer->name_length);
     if (!msg->via.grp.author) {
         debug_error("Groupchat:\tUnable to allocate space for author nickname.\n");
+
+        free(msg);
+        return UINT32_MAX;
     }
+    memcpy(msg->via.grp.author, peer->name, peer->name_length);
 
     msg->via.grp.msg = calloc(1, length);
     if (!msg->via.grp.msg) {
         debug_error("Groupchat:\tUnable to allocate space for message.\n");
-    }
 
-    memcpy(msg->via.grp.author, nick,    peer->name_length);
-    memcpy(msg->via.grp.msg,    message, length);
+        free(msg->via.grp.author);
+        free(msg);
+        return UINT32_MAX;
+    }
+    memcpy(msg->via.grp.msg, message, length);
 
     pthread_mutex_unlock(&messages_lock);
+
     return message_add_group(m, msg);
 }
 
@@ -94,9 +101,15 @@ void group_peer_add(GROUPCHAT *g, uint32_t peer_id, bool UNUSED(our_peer_number)
 
     GROUP_PEER *peer = g->peer[peer_id];
 
+    char *default_peer_name = "<unknown>";
+
     // Allocate space for the struct and the dynamic array holding the peer's name.
-    peer = calloc(1, sizeof(GROUP_PEER) + 10); // 10 is strlen("<unknown>") + \0
-    strcpy2(peer->name, "<unknown>");
+    peer = calloc(1, sizeof(GROUP_PEER) + strlen(default_peer_name) + 1);
+    if (!peer) {
+        debug_error("Groupchat:\tUnable to allocate space for group peer.\n");
+        exit(42); // TODO: Header file for exit codes. This is just silly.
+    }
+    strcpy2(peer->name, default_peer_name);
     peer->name_length = 0;
     peer->name_color  = name_color;
     peer->id          = peer_id;
@@ -141,12 +154,12 @@ void group_peer_name_change(GROUPCHAT *g, uint32_t peer_id, const uint8_t *name,
     if (peer && peer->name_length) {
         uint8_t old[TOX_MAX_NAME_LENGTH];
         uint8_t msg[TOX_MAX_NAME_LENGTH];
-        size_t  size = 0;
+        size_t size = 0;
 
         memcpy(old, peer->name, peer->name_length);
-        size = snprintf((char *)msg, TOX_MAX_NAME_LENGTH, "<- has changed their name from %.*s", (int)peer->name_length,
-                        old);
-        peer = realloc(peer, sizeof(GROUP_PEER) + sizeof(void) * length);
+        size = snprintf((char *)msg, TOX_MAX_NAME_LENGTH, "<- has changed their name from %.*s",
+                        (int)peer->name_length, old);
+        peer = realloc(peer, sizeof(GROUP_PEER) + sizeof(char) * length);
 
         if (peer) {
             peer->name_length = utf8_validate(name, length);
@@ -156,13 +169,12 @@ void group_peer_name_change(GROUPCHAT *g, uint32_t peer_id, const uint8_t *name,
             group_add_message(g, peer_id, msg, size, MSG_TYPE_NOTICE);
             return;
         } else {
-            debug("Fatal error:\t couldn't realloc for group peer name!\n");
+            debug_error("Groupchat:\t couldn't realloc for group peer name!\n");
             exit(40);
         }
-
     } else if (peer) {
         /* Hopefully, they just joined, because that's the UX message we're going with! */
-        peer = realloc(peer, sizeof(GROUP_PEER) + sizeof(void) * length);
+        peer = realloc(peer, sizeof(GROUP_PEER) + sizeof(char) * length);
         if (peer) {
             peer->name_length = utf8_validate(name, length);
             memcpy(peer->name, name, length);
@@ -172,7 +184,7 @@ void group_peer_name_change(GROUPCHAT *g, uint32_t peer_id, const uint8_t *name,
             return;
         }
     } else {
-        debug("Fatal error:\t we can't set a name for a null peer! %u\n", peer_id);
+        debug_error("Groupchat:\tWe can't set a name for a null peer! %u\n", peer_id);
         exit(41);
     }
 }
@@ -197,6 +209,8 @@ void group_free(GROUPCHAT *g) {
     group_reset_peerlist(g);
 
     for (size_t i = 0; i < g->msg.number; ++i) {
+        free(g->msg.data[i]->via.grp.author);
+        free(g->msg.data[i]->via.grp.msg);
         message_free(g->msg.data[i]);
     }
     free(g->msg.data);

--- a/src/groups.h
+++ b/src/groups.h
@@ -61,6 +61,7 @@ GROUPCHAT group[64];
 
 void group_init(GROUPCHAT *g, uint32_t group_number, bool av_group);
 
+// Returns the message number on success, returns UINT32_MAX on failure.
 uint32_t group_add_message(GROUPCHAT *g, uint32_t peer_id, const uint8_t *message, size_t length, uint8_t m_type);
 
 void group_peer_add(GROUPCHAT *g, uint32_t peer_id, bool our_peer_number, uint32_t name_color);

--- a/src/messages.c
+++ b/src/messages.c
@@ -195,6 +195,11 @@ static bool msg_add_day_notice(MESSAGES *m, time_t last, time_t next) {
         || (ltime_year == msg_time->tm_year && ltime_mon == msg_time->tm_mon && ltime_day < msg_time->tm_mday))
     {
         MSG_HEADER *msg = calloc(1, sizeof(MSG_HEADER));
+        if (!msg) {
+            debug_error("Messages:\tCouldn't allocate memory for day notice.\n");
+            return false;
+        }
+
         time(&msg->time);
         msg->our_msg       = 0;
         msg->msg_type      = MSG_TYPE_NOTICE_DAY_CHANGE;
@@ -286,6 +291,10 @@ uint32_t message_add_type_action(MESSAGES *m, bool auth, const char *msgtxt, uin
 
 uint32_t message_add_type_notice(MESSAGES *m, const char *msgtxt, uint16_t length, bool log) {
     MSG_HEADER *msg = calloc(1, sizeof(MSG_HEADER));
+    if (!msg) {
+        debug_error("Messages:\tCouldn't allocate memory for notice.\n");
+        return UINT32_MAX;
+    }
 
     time(&msg->time);
     msg->our_msg       = 0;
@@ -885,7 +894,7 @@ static int messages_draw_group(MESSAGES *m, MSG_HEADER *msg, uint32_t curr_msg_i
         h2 = UINT32_MAX;
     }
 
-    messages_draw_author(x, y, MESSAGES_X - NAME_OFFSET, msg->via.grp.msg, msg->via.grp.author_length, msg->via.grp.author_color);
+    messages_draw_author(x, y, MESSAGES_X - NAME_OFFSET, msg->via.grp.author, msg->via.grp.author_length, msg->via.grp.author_color);
     messages_draw_timestamp(x + width, y, &msg->time);
     return messages_draw_text(msg->via.grp.msg, msg->via.grp.length, msg->height, msg->msg_type, msg->our_msg, 1,
                               h1, h2, x + MESSAGES_X, y, width - TIME_WIDTH - MESSAGES_X, height)

